### ISSUE TITLE
[Feature]: Add dark theme support to Jaeger-UI

### ISSUE THREAD

**Issue by anshgoyalevil** [jaegertracing/jaeger-ui#1911]
### Requirement

Since Jaeger is a distributed tracing platform, we sometimes need to look at the traces the whole day. Working with a high-contrast UI becomes a pain for the eye in such cases.

### Problem

Jaeger-UI currently supports only the light theme. A dark theme would help users focus on observability rather than eye strain.

### Proposal

Since react v18, and ant-design v5, the support for dynamic theming and state management has improved.

ant-design provides three theme algorithms: `defaultAlgorithm`, `darkAlgorithm`, and `compactAlgorithm`. Our concern is with the first two.

We can change the `jaegerTheme`'s default algorithm based on the current state of the theme. For example:

```diff
const jaegerTheme = {
+ algorithm: theme.darkAlgorithm,
  token: {
    ...defaultTheme.token,
    colorPrimary: '#199',
  },
  components: {
    ...defaultTheme.components,
    Layout: {
      ...defaultTheme.components.Layout,
      bodyBg: '#fff',
      headerBg: '#404040',
      footerBg: '#fff',
      headerHeight: 48,
      headerPadding: '0 50',
....rest of code
```
will change the whole Jaeger-UI to dark mode, but currently, since we haven't designed the whole UI that way, there are many areas where the theme still remains the same, i.e., light.

State management features of react (ex, `useState` in react 18) can be used to manage the current theme's value and can be used to set the theme dynamically based on the click of a button, like in most applications.

Actionable items:
- Remove the hard-coded themes from the codebase, and make use of colors based on the current theme state's algorithm
- Go step-by-step, with small code changes/PRs at each step.
- Use statement management to change/store the current theme.
- Add a Switch kind of component to the NavBar to toggle between dark and light modes.

Relevant documentation:
- ant-design theme customization guide: https://ant.design/docs/react/customize-theme
- https://ant.design/docs/spec/dark

**Comment by yurishkuro** [comment]
This is not a small issue in itself, but can be broken into many small changes that are actually `good-first-issue`s

**Comment by yurishkuro** [comment]
@anshgoyalevil [this section](https://ant.design/docs/react/customize-theme#seedtoken) describes a lot of pre-defined names/tokens. Is the intention that our higher-level components should be using those tokens for their styles as well? We have a bunch of custom styles, I am wondering if they all need to be migrated, or if the approach that we can keep those custom tokens but migrate them into the themes framework to allow switching.

**Comment by anshgoyalevil** [comment]
Yes. Most probably, we would have to create the two themes, light, and dark. Both have their own algorithms, and design tokens according to the style we want to apply to that specific component. The `useToken()` hook would allow us to consume the design tokens specified in the theme object.

For example:
```tsx
const darkTheme = {
  algorithm: theme.darkAlgorithm,
  token: {
    ...defaultTheme.token,
    colorPrimary: '#fff', // Primary color for dark mode
  },
...rest of the theme object
```

```tsx
const lightTheme = {
  algorithm: theme.defaultAlgorithm,
  token: {
    ...defaultTheme.token,
    colorPrimary: '#000', // Primary color for light mode
  },
...rest of the theme object
```
We can toggle between these two themes using the useState hook, or any other such thing. Now to consume the values defined over here, we can do it using the useToken hook like `useToken().colorPrimary` etc.
This useToken() hook can be consumed via the props received from `withRouteProps` HOC, or any other such approach because we are using class components and can't directly use hooks inside them.

So yes, that would be a lot of work to do, since we are trying to introduce dark mode support to a large codebase.

**Comment by AHB102** [comment]
@yurishkuro So is this good for someone like me ? I am new to open source, would like to contribute but don't know how to, if so, can I start contributing ?

**Comment by yurishkuro** [comment]
you're welcome to try. For general OSS contributing background see https://contribute.cncf.io/contributors/getting-started

**Comment by zacanger** [comment]
I found this issue because I'm looking at customizing the theme. Two things that might be useful:

* Allowing end-user theme config (merging with or overriding the default theme)
* A hook to set dark mode based on user preferences, something like:

```javascript
  const [darkMode, setDark] = useState(false)
  useEffect(() => {
    const useDark = window.matchMedia('(prefers-color-scheme:dark)').matches
    setDark(useDark)
    return
  })
  ```

**Comment by create2000** [comment]
Hi, My name is Anthony. I am applying for the LFX Mentorship and i chose this project. is this issue still open?

**Comment by yurishkuro** [comment]
yes it is

**Comment by muskan-rathor** [comment]
Hi! I'd love to work on this issue as my first contribution to Jaeger UI. I understand this is a large task, so I'd like to start by working on a small part, like introducing the `enableThemes` config flag and basic dark/light toggle. Please let me know if that’s okay, and I’ll get started.

**Comment by yurishkuro** [comment]
Ok

**Comment by adity-a34** [comment]
Hi @yurishkuro @anshgoyalevil 
As per the repository, this issue has not been implemented.
I have this implementation plan:

1. Introduce Theme Toggle Config
- Add an enableThemes flag to the config (default: false).
- Update config files (e.g., default-config.tsx) and ensure the flag is accessible in the app.

2. Define Theme Objects
- Create lightTheme and darkTheme objects using Ant Design’s defaultAlgorithm and darkAlgorithm.
- Set up design tokens for both themes.

3. Implement Theme State Management
- Use React’s useState (or context if needed) to manage the current theme.
- Default to light theme; allow toggling to dark.

4. Add Theme Switch UI
- Add a Switch component to the NavBar (visible only if enableThemes is true).
- On toggle, update the theme state.

5. Apply Theme Dynamically
- Pass the current theme object to Ant Design’s ConfigProvider.
- Ensure the theme updates throughout the app when toggled.

6. Refactor Custom Styles
- Gradually migrate custom styles to use theme tokens (via useToken() or props).
- Replace hardcoded colors with theme-based values.

7. Testing and Documentation

Please tell me if I should proceed with this implementation or if this issue is under implementation by someone else.
Also, please correct my implementation if I should introduce any changes.

**Comment by yurishkuro** [comment]
@ADITYATIWARI342005 i have the intuition that adding themes is going to be a lot of work. I don't know how much work, as I am not a Ui developer and have no experience with theming. So before starting any of the work I want to understand the scope / effort and the risks. Your plan does not address these questions, so it's not helpful to me at this stage. In fact, your plan starts with the trivial, the least risky tasks like adding a selector - a certain amount of work that in no way addresses the unknown risks or improves the chances of success.

**Comment by adity-a34** [comment]
@yurishkuro 
Thanks for the feedback and for calling out the gaps. Indeed, this would take a significant amount of time and effort, and there is a major risk of incomplete coverage and issues in future maintenance due to consideration of both themes. 
But it also has a remarkable impact if implemented successfully, which was the reason for my interest in giving it a try. 
But analyzing the risk of inconsistent themes leading to poor readability and the introduction of unintentional bugs/performance issues, it is better to be implemented under a maintainer/mentor's guidance so as to mitigate these challenges.

**Comment by axsaucedo** [comment]
+1 this would be great!

@yurishkuro kudos on the progress, looks promising!

QQ, what is the current roadmap? If shared as suggested as various good-first-issues perhaps indeed other devs can pick these up? 

(I can at least give some a shot as well)

**Comment by yurishkuro** [comment]
the bulk of work is already done, there are some here and there glitches that are being fixed by contributors (the bottleneck is in code reviews not in the implementations).

**Comment by axsaucedo** [comment]
Great! Thank you @yurishkuro
